### PR TITLE
Dispatch the ``partd.Encode`` class used for disk-based shuffling

### PR DESCRIPTION
**Pull request by rjzamora**
This change allows dask-cudf to register the necessary dispatch function (`partd_encode_dispatch`) for cudf-backed data to work with `shuffle="disk"`. Right now, the disk-based shuffle is hardcoded to use `partd.PandasBlocks` (a subclass of `partd.Encode`).

### CODE REVIEW

**Comment by rjzamora** [comment]
I'd consider this change to be "low risk." So, I'm planning to merge EOD if there are no comments.

**Review by wence- — APPROVED** [review]
Some very minor nits

File: dask/dataframe/shuffle.py [hunk 1]
```diff
@@ -577,7 +589,8 @@ def rearrange_by_column_disk(df, column, npartitions=None, compute=False):
     always_new_token = uuid.uuid1().hex
 
     p = ("zpartd-" + always_new_token,)
-    dsk1 = {p: (maybe_buffered_partd(),)}
+    encode_cls = partd_encode_dispatch(df._meta)
```
[the remark this answers is not in the document]
Comment (rjzamora) [reply dask/dataframe/shuffle.py:592]: >or does that happen automagically once the on-disk partd encoder is different?

Yeah, the `Encode` class includes both the `serialize` and `deserialize` methods needed to encode/decode.

File: dask/dataframe/shuffle.py [hunk 2]
```diff
@@ -521,16 +525,23 @@ class maybe_buffered_partd:
...
+            return (maybe_buffered_partd, (self.encode_cls, False, self.tempdir))
         else:
-            return (maybe_buffered_partd, (False,))
+            return (
+                maybe_buffered_partd,
+                (
+                    self.encode_cls,
+                    False,
+                ),
+            )
```
[the remark this answers is not in the document]
Comment (rjzamora) [reply dask/dataframe/shuffle.py:544]: Oh weird! Wonder when that happened.

**Review by wence- — APPROVED** [review]

File: dask/dataframe/shuffle.py [hunk 3]
```diff
@@ -521,16 +525,23 @@ class maybe_buffered_partd:
     If serialized, will return non-buffered partd. Otherwise returns a buffered partd
     """
 
-    def __init__(self, buffer=True, tempdir=None):
+    def __init__(self, encode_cls=None, buffer=True, tempdir=None):
         self.tempdir = tempdir or config.get("temporary_directory", None)
         self.buffer = buffer
         self.compression = config.get("dataframe.shuffle.compression", None)
+        self.encode_cls = encode_cls  # Defaults to partd.PandasBlocks
```
[the remark this answers is not in the document]
Comment (rjzamora) [reply dask/dataframe/shuffle.py:532]: Yeah, we will need to add a partd import here, but it's probably best to set the default earlier.

**Pull request merged**